1-Draft/RFCNNNN-Dispose-Script-Block.md

```diff
@@ -0,0 +1,242 @@
+
+- Error
+- Warning
+- Verbose
+- Debug
+- Information
+
+#### Error States
+
+Terminating errors can also be thrown from a `dispose{}` block, which will skip the invocation of the rest of that `dispose{}` block, without preventing other `dispose{}` blocks from being executed.
+
+### Implementation Details
+
+- A new `dispose` keyword, indicating a `NamedBlockAst` similarly to `begin`, `process`, and `end`, which is valid in the same contexts as the existing named block keywords.
+  - This includes new token types and appropriate entries in the `Tokenizer`'s keyword tables.
+- Additional members and constructors for `ScriptBlockAst` to enable it to be recognised by the parser.
+- Additional methods in `DlrScriptCommandProcessor` and `CompiledScriptBlock` to enable the block to be executed at the correct time.
+  - Both types implement or inherit from a type that implements `IDisposable`, and as such the existing `Dispose()` methods can generally be utilised to run this block.
+- Changes to the behaviour of command execution in `CommandProcessorBase` to recognise when a function or script cmdlet is being executed, and appropriately call their `Dispose()` methods to invoke the dispose block.
+  - This can include changes to the existing command disposal behaviour, to enable commands to dispose their resources as soon as their tasks in the pipeline are completed, instead of waiting for the entire pipeline to complete.
+  - Also required with this is an additional layer of `try/finally` during `Dispose()` events as we are opening up the possibility for users to throw terminating errors _during_ a command's disposal, and as such we must ensure that disposal still completes appropriately.
+
+#### `Cmdlet` / `PSCmdlet`
+
+Commands inheriting from `Cmdlet` or `PSCmdlet` can already implement a version of this behaviour.
+To do so, such commands need also implement `IDisposable` and the `Dispose()` method.
+This RFC would simply change when the dispose method is called.
+Currently, the method is called during pipeline disposal.
+This RFC would ensure it is called immediately after `EndProcessing()` completes during normal pipeline operation.
+It will continue to be called in the case of a terminating error as it is currently.
+
+#### `ForEach-Object` Additions
+
+- `ForEachObjectCommand` should implement `IDisposable` to enable similar behaviour for the cmdlet.
+- A `-DisposeBlock` parameter can be exposed to allow it to be called from scripts.
+  - The slightly different parameter name is necessary as C# does not permit a property name to collide with a method name, and `IDisposable` mandates a method named `Dispose()`.
+  - To improve consistency, the parameter can be explicitly aliased to `-Dispose`.
+  - The parameter _must_ be assigned explicitly by name, and never be assumed based on number of provided scriptblock parameters as the command currently does with `begin`/`process`/`end`.
+
```

Comment:
I'm not sure. If I don't, `dispose {}` is basically a dead block there, I don't see a _good_ way to prevent folks using it.

In truth, wasn't really aware they supported named blocks till the call today. Open to suggestions on this one, really could go either way. Kind of like `ForEach-Object` there. I've actually moved this section to Considerations as well. 🙂